### ISSUE TITLE
[BUG] Checkboxes with an initial value of "true" will be checked every time the modal is opened

### ISSUE THREAD

**Issue by Pamasich** [aclist/kbin-kes#248]
**Describe the bug**
When an add-on gives a checkbox an initial value of `true`, the checkbox will be in a checked state every time the modal is opened. The underlying setting still only flips with actual human clicks on the checkbox, but the checkbox's state doesn't honor what the setting currently is. This only happens if the initial value is true, not with a `false` value.

**To Reproduce**
1. Make sure the `Hide sidebar entirely` checkbox of the `Hide sidebar elements` add-on is unchecked.
2. Change the `manifest.json` (in Tampermonkey via KES's script page -> Externals -> Resources -> Edit) by adding a new line `"initial": true` to the `Hide sidebar entirely` setting.
3. Save and reload kbin
4. The sidebar is still there, but opening the KES settings modal will have the option to hide it checked.
5. Uncheck the checkbox, close the KES settings, then open them again — the checkbox is automatically checked again

I'm just using the `Hide sidebar elements` add-on as an example here, any checkbox has the issue.

**Expected behavior**
The checkbox should reflect the state of the underlying setting it represents. The `initial` value set by the add-on should be the default in cases where the user hasn't actively selected anything yet. Not something that determines the state of the checkbox, independently of the setting, every time the modal is opened.

**Screenshots**
[Screenshot of the Hide sidebar entirely checkbox displaying as turned on, but the sidebar is still visible in the background.](https://files.catbox.moe/8qfoi0.png)

**KES version:**
- `testing` branch, on commit `aadc52383be90052cd878199bfa53eebfd2a38b4` (latest as of writing this)

**Extension and version:**
- Extension: Tampermonkey
- Version number: 5.0.1 (latest)

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser: Edge
 - Browser version: 121.0.2277.83

**Comment by Pamasich** [comment]
I've noticed that, in addition, when the page is newly loaded the setting gets turned on automatically. This is different from what I reported above where opening the modal has the checkbox automatically checked irrespective of the underlying setting. In that case, only the checkbox is affected, the setting doesn't get turned on. But when reloading or loading a new page, the setting does actually reset to true.

**Comment by aclist** [comment]
I've been thinking about this for a few days and looking at the relevant code, but was unable to yet find a situation where this would actually occur in real-world use. Here's a breakdown of how KES loads settings:

1. Populate the menu with items from the manifest
2. If a particular mod is already saved as ON in the user's local config, apply it. Otherwise, start off.
3. If a particular mod's sub-settings have values saved in the user's local config, apply those accordingly (e.g. colors in color pickers, state of radio buttons and checkboxes, etc.). Otherwise, for visual purposes, set them to the initial state, if applicable (required for some field types), or to empty/off. N.B. Checkboxes were never required to supply an initial state in the manifest, being seen as optional sub-settings.
4. Upon installing KES for the first time or downloading a version update with new mods, such new mods are by definition set to OFF because the user has never enabled them before. Therefore, the state of any sub-settings or menus within the mod (colors, labels, dropdown selections, checkbox toggles) will be cosmetically displayed in the KES modal only, but such settings are not saved to the user's local config, since these represent only default settings.
5. Once the user actively toggles the mod to ON, the settings shown within the sub-menu at that time are saved into the local config and propagated on the page.

Given the above, if a mod sets the `initial` value of a checkbox to `true` and the checkbox is displayed, but the mod has never been turned ON, this setting will not be applied. As soon as the user sets the mod to ON, the setting represented by the `true` checkbox will be saved to the configs and also applied.

Therefore, any mod shipping with checkboxes in the initial state of true will only apply those settings after the user manually invokes the toggle to ON for the whole mod.

I think the issue in your test above is that you are editing the manifest for a mod that is already "hot," i.e., the mod is already ON, so new settings aren't being applied when you added new parameters to the manifest. I don't think this is a real-world scenario, since a new manifest would only be coming if the user updated KES, and if it's a new, never-before-enabled mod, its settings would indeed take effect once the user toggled it on, inclusive of any checkboxes set to an initial state of `true`.

The only use case I could see for this is if:

1. A mod that previously shipped adds a new setting
2. The setting is given an initial value of `true`
3. The user already had the mod toggled ON prior to the update

Outcome: the new setting won't be applied until the user toggles the mod OFF and then ON again.

This does seem like a possible pitfall that should be corrected, since it's not entirely impossible for a mod to ship with later updates that introduce new sub-settings. Traditionally we dealt with this by reminding the user to pulse the mod off and on again.

There may also be some logic that reapplies the settings on page reload, but I can't recall at the moment. The above should be the expected control flow.

When mocking changes to the manifest or other parts of KES, I suggest updating the grants/remote URLs and using something like a [gist](https://gist.github.com/) to quickly host the file (might take a few seconds to cache on GitHub's side), rather than updating *Monkey's locally-stored files, which might produce false positives.

I think this answers your question, but let me know if it doesn't.

**Comment by Pamasich** [comment]
I originally encountered this issue while testing #247 with a custom [kes.user.json](https://github.com/Pamasich/kbin-kes/blob/test/checkbox-bug/kes.user.js). I just used the monkey's manifest here because I didn't think of the option of changing the remote url while writing this and figured it would have the same effect.

I took a look at the code with the browser's dev tools and the issue occurs [on line 578](https://github.com/aclist/kbin-kes/blob/main/kes.user.js#L578):

```javascript
if (!modSettings[key]) {
    modSettings[key] = initial;
    saveModSettings(modSettings, ns);
}
```

An unchecked checkbox is `false`, so this code gets triggered and resets it to `true` whenever the KES menu is opened. Even if it's been manually turned off before.
